docs/examples/platform/node_helpers.hpp

```diff
@@ -0,0 +1,294 @@
+
+    template <typename T, typename TxSession, typename TxMetadata>
+    static cetl::optional<libcyphal::transport::AnyFailure> serializeAndSend(const T&          value,
+                                                                             TxSession&        tx_session,
+                                                                             const TxMetadata& metadata)
+    {
+        using traits = typename T::_traits_;
+        std::array<std::uint8_t, traits::SerializationBufferSizeBytes> buffer{};
+
+        const auto data_size = serialize(value, buffer).value();
+
+        // NOLINTNEXTLINE
+        const cetl::span<const cetl::byte> fragment{reinterpret_cast<cetl::byte*>(buffer.data()), data_size};
+        const std::array<const cetl::span<const cetl::byte>, 1> payload{fragment};
+
+        return tx_session.send(metadata, payload);
+    }
+
+    template <typename T>
+    static bool tryDeserialize(T& obj, const libcyphal::transport::ScatteredBuffer& buffer)
+    {
+        std::vector<std::uint8_t>             data_vec(buffer.size());
+        const auto                            data_size = buffer.copy(0, data_vec.data(), data_vec.size());
+        const nunavut::support::const_bitspan bitspan{data_vec.data(), data_size};
+
+        return deserialize(obj, bitspan);
+    }
+
+    template <typename Message>
+    static cetl::optional<Publisher<Message>> makeAnyPublisher(Presentation&                      presentation,
+                                                               const libcyphal::transport::PortId subject_id)
+    {
+        auto maybe_publisher = presentation.makePublisher<Message>(subject_id);
+        EXPECT_THAT(maybe_publisher, testing::VariantWith<Publisher<Message>>(testing::_))
+            << "Failed to create publisher (subject_id=" << subject_id << ").";
+        if (auto* const publisher = cetl::get_if<Publisher<Message>>(&maybe_publisher))
+        {
+            return std::move(*publisher);
+        }
+        return cetl::nullopt;
+    }
+
+    template <typename Message>
+    static cetl::optional<Subscriber<Message>> makeAnySubscriber(Presentation&                      presentation,
+                                                                 const libcyphal::transport::PortId subject_id)
+    {
+        auto maybe_subscriber = presentation.makeSubscriber<Message>(subject_id);
+        EXPECT_THAT(maybe_subscriber, testing::VariantWith<Subscriber<Message>>(testing::_))
+            << "Failed to create subscriber (subject_id=" << subject_id << ").";
+        if (auto* const subscriber = cetl::get_if<Subscriber<Message>>(&maybe_subscriber))
+        {
+            return std::move(*subscriber);
+        }
+        return cetl::nullopt;
+    }
+
+    struct Heartbeat
+    {
+        using Message = uavcan::node::Heartbeat_1_0;
+
+        bool makeRxSession(libcyphal::transport::ITransport&                                      transport,
```

Comment:
You are right that it's not needed for presentation layer, but the same `NodeHelpers::xxx` stuff is in use for transport layer examples (02 & 03). Pavel mentioned that he wants more self-sufficient examples (and I hear you guys, and I actually started like this initially), but I really struggle with all this massive code duplication I need to introduce and maintain if made each and every example self-sufficient (aka 1 single `.cpp` file)... and there will be more example apps... With `NodeHelpers` I can avoid this.

Here is an idea: what if I will add one intermediate layer-related subdirectory into `docs/examples` hierarchy, so that they will be groped by "0_transport" and "1_presentation" (and "2_application"), and more importantly each of such layer-related dirs will contain only its own set of `NodeHelpers` with stuff which is boilerplate/specific to this layer only. Something like this:
```
docs /
    dsdl
    platform
    0_transport
        example_02_posix_udp
        example_03_linux_socketcan
    1_presentation
        example_12_posix_udp        
```

?

Reply:
Demos/examples should not rely on helpers at all except for platform-specific parts, because they reduce the clarity of the example. If there is too much code duplication, we should look into altering the API design.